### PR TITLE
Implement Joining and leaving mission action

### LINKED ISSUE #9: [1pt] Implement mission leaving - Actions
- Follow the same logic as with the "Join mission" - but you need to set the `reserved` key to `false`. 
- Dispatch these actions upon click on the corresponding buttons.

### LINKED ISSUE #10: [3pt] Implement mission joining - Actions
- When a user clicks the "Join Mission" button, action needs to be dispatched to update the store. You need to get the ID of the selected mission and update the state. Remember you mustn't mutate the state. Instead, you need to return a new state object with all missions, but the selected mission will have an extra key `reserved` with its value set to `true`. You could use a JS `filter()` or `map()` to set the value of the new state - i.e.:
```javascript
const newState = state.map(rocket => {
    if(mission.id !== id) 
        return mission;
    return { ...mission, reserved: true };
});

```

- Regardless of which method you choose, make sure you place all your logic in the reducer. In the React view file, you should only dispatch the action with the correct rocket ID as an argument.

### PR DESCRIPTION
**Pull request by desorgui**
**Changes**
- Implement Joining mission action
- Implement leaving mission action

### CODE REVIEW

**Review by HansZizold — APPROVED** [review]
Hi @desorgui,

I have reviewed your code. I don't have any observation. So you can merge it.
Nice job!!!

**Comment by desorgui** [comment]
Thank you @HansZizold.

File: src/Components/Missions.js [hunk 1]
```diff
@@ -1,8 +1,36 @@
...
+            <th />
+          </tr>
+          {missions &&  missions.map((mission) => (
+          <tr key={mission.mission_id}>
+            <td className="missionName">{mission.mission_name}</td>
+            <td className="missionDescription">{mission.description}</td>
+            <td className={`userBadge ${mission.reserved ? 'member' : ' '}`}><label>
+              {mission.reserved ? 'Active member' : 'NOT A MEMBER'}</label></td>
+            <td className="joinButton">
+              <button onClick={() => dispatch(joinMission(mission.mission_id))} className={`missionBtn ${mission.reserved ? 'joined' : ' '}`}>
```
Comment (HansZizold) [inline src/Components/Missions.js:25]: In this line you're creating a new function in every render which can be less performant. You can improve this by moving the function outside the return statement like this:

const handleJoinMission = (missionId) => {
    dispatch(joinMission(missionId));
}
Then, you can use this function inside the onClick event:

<button onClick={() => handleJoinMission(mission.mission_id)} className={`missionBtn ${mission.reserved ? 'joined' : ' '}`}>

File: src/Components/Missions.js [hunk 2]
```diff
@@ -1,8 +1,36 @@
...
+            <th>Description</th>
+            <th>Status</th>
+            <th />
+          </tr>
+          {missions &&  missions.map((mission) => (
+          <tr key={mission.mission_id}>
+            <td className="missionName">{mission.mission_name}</td>
+            <td className="missionDescription">{mission.description}</td>
+            <td className={`userBadge ${mission.reserved ? 'member' : ' '}`}><label>
+              {mission.reserved ? 'Active member' : 'NOT A MEMBER'}</label></td>
```
Comment (HansZizold) [inline src/Components/Missions.js:23]: Semantic HTML: you are using a label element to display the membership status. However, label elements are typically used for form input descriptions. Consider replacing this with a more semantic HTML element such as p or span.

File: src/redux/missions/missions.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        missionArr.push({
+            mission_id: mission.mission_id,
+            mission_name: mission.mission_name,
+            description: mission.description,
+            reserved: false
+        });
+      }));
+      return (missionArr);
+    },
+  );
```
Comment (HansZizold) [inline src/redux/missions/missions.js:23]: The getMissions thunk could be improved for readability. The use of async/await syntax mixed with .then syntax is not typical and could be confusing. You should consider using just async/await like this:

export const getMissions = createAsyncThunk(
    'missions/fetchMissions',  // String for the action type
    async () => {
        const response = await fetch(getUrl);
        const missionsData = await response.json();
        const missionArr = missionsData.map(mission => ({
            mission_id: mission.mission_id,
            mission_name: mission.mission_name,
            description: mission.description,
            reserved: false
        }));
        return missionArr;
    },
);

File: src/redux/missions/missions.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
+
+const getUrl = 'https://api.spacexdata.com/v3/missions';
+
+const FETCHED_MISSION = 'FETCHED_MISSION';
+const JOIN_OR_LEAVE_MISSION = 'JOIN_OR_LEAVE_MISSION';
```
Comment (HansZizold) [inline src/redux/missions/missions.js:6]: There's no need to declare these constant action types (FETCHED_MISSION and JOIN_OR_LEAVE_MISSION) when using createAsyncThunk and createSlice. The advantage of Redux Toolkit is that it automatically generates action types. So for example in line 9 you can do:

'missions/fetchMissions', // String for the action type

File: src/redux/missions/missions.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    },
+  );
+  export const joinMission = (id) => ({ type: JOIN_OR_LEAVE_MISSION, payload: id });
+
+  const missionSlice = createSlice({
+    name: 'missions',
+    initialState: [],
+    extraReducers: {
+      [getMissions.fulfilled]: (state, action) => action.payload,
+      [JOIN_OR_LEAVE_MISSION]: (state, action) => (
```
Comment (HansZizold) [inline src/redux/missions/missions.js:31]: You can use the auto-generated action creators for 'joinMission' action:

joinOrLeaveMission: (state, action) => { ... }

**Review by HansZizold — COMMENTED** [review]
# Changes required :recycle: 

Please check my comments below and apply the required changes.
Submit for another review when you finish.

Thanks and happy coding :grin:

**Comment by HansZizold** [comment]
Hi Desor, 

I did a new review of our project, now applying new knowledge I learned on my coding way :wink: .

Thank you, best regards,

**Pull request merged**